Detox (E2E) test automation (#3199)
codemagic-ci-cd/codemagic-docs: content/yaml-testing/detox-testing.md

---
title: Detox (E2E) test automation
description: How to run Detox tests with codemagic.yaml
weight: 5

---

Detox testing refers to end-to-end (E2E) testing for React Native apps using the Detox framework. Detox provides a gray box testing solution, meaning it runs tests on a real device or emulator while also accessing the app's internal state for better synchronization.

## Prerequisites

1. A React Native project
2. **npm/yarn** are pre-installed
3. Xcode for iOS and Android Studio for Android are pre-installed
4. **homebrew** is pre-installed

{{<notebox>}}
Pre-installed means that Codemagic machines already have the software ready to use, so there is no need to install it manually.
{{</notebox>}}

## Running Detox tests from codemagic.yaml

> 1. Adjust your project's **package.json** file with Detox before starting Codemagic configurations:

Author: yea, that's what I thought and added "before starting Codemagic configurations:" but now thinking if i just remove the steps and use headers instead.

```json
  "detox": {
    "runnerConfig": "e2e/config.json",
    "configurations": {
      "ios.sim.release": {
        "type": "ios.simulator",
        "device": {
          "type": "iPhone 14"
        },
        "binaryPath": "ios/build/Build/Products/Debug-iphonesimulator/MyReactNativeApp.app",
        "build": "xcodebuild -workspace ios/YOUR_APP.xcworkspace -scheme YOUR_APP -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build"

      },
    }
  }
```

> 2. Install the required tools by running the following commands:

Reviewer: "Running Detox tests in Codemagic"

We are going to add all the scripts we need to run to install additional software and execute Detox tests to the `scripts` section of codemagic.yaml.

---
title: Detox (E2E) test automation
description: How to run Detox tests with codemagic.yaml
weight: 5

---

Detox testing refers to end-to-end (E2E) testing for React Native apps using the Detox framework. Detox provides a gray box testing solution, meaning it runs tests on a real device or emulator while also accessing the app's internal state for better synchronization.

## Prerequisites

1. **npm/yarn** are pre-installed

> 2. A ready React Native project

Reviewer: Maybe you should start the list with React Native project here? Also, what do you mean by ready? Maybe simply "A React Native project"

3. Xcode for iOS and Android Studio for Android are pre-installed

> 4. **homebrew** are pre-installed

Reviewer: ```suggestion
4. **homebrew** is pre-installed
```

{{<notebox>}}

> Pre-installed means that Codemagic machines already have them ready to use, so no need to install them manually.

Reviewer: ```suggestion
Pre-installed means that Codemagic machines already have the software ready to use, so there is no need to install it manually.
```

{{</notebox>}}

## Running Detox tests

1. Configure your project's **package.json** file with Detox:

```json
  "detox": {
    "runnerConfig": "e2e/config.json",
    "configurations": {
      "ios.sim.release": {
        "type": "ios.simulator",
        "device": {
          "type": "iPhone 14"
        },
        "binaryPath": "ios/build/Build/Products/Debug-iphonesimulator/MyReactNativeApp.app",
        "build": "xcodebuild -workspace ios/YOUR_APP.xcworkspace -scheme YOUR_APP -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build"

      },
    }
  }
```
2. Install **detox-cli** by running **npm install detox-cli --global**

3. Add a script to install **applesimutils** as it is required by Detox to work with iOS simulators:

```yaml
brew tap wix/brew
brew install applesimutils
```

4. To build the app, run the following command:

```yaml
detox build --configuration ios.sim.release
```

> 5. Once you successfully installed the necessary tools and built your app, it is time to test it:

Reviewer: At first I read it as individual instructions that need to be carried out, maybe it's better to combine them and start by saying the user must now now edit the codemagic.yaml file, and then present the sample yaml file.

{{</notebox>}}

## Preparing the application

Adjust your project's **package.json** file with Detox before starting Codemagic configurations:

```json
  "detox": {
    "runnerConfig": "e2e/config.json",
    "configurations": {
      "ios.sim.release": {
        "type": "ios.simulator",
        "device": {
          "type": "iPhone 14"
        },
        "binaryPath": "ios/build/Build/Products/Debug-iphonesimulator/MyReactNativeApp.app",
        "build": "xcodebuild -workspace ios/YOUR_APP.xcworkspace -scheme YOUR_APP -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build"

      },
    }
  }
```

More information about Detox environment setup can be found in the official documentation [here](https://wix.github.io/Detox/docs/introduction/environment-setup).

## Running Detox tests in Codemagic

We are going to add all the scripts we need to run to install additional software and execute Detox tests to the `scripts` section of `codemagic.yaml`.

1. Install Detox CLI tools and **applesimutils** which is required by Detox to work with iOS simulators:
   
```bash
npm install detox-cli --global
brew tap wix/brew
brew install applesimutils
```

> 2. Build and run detox tests:

Reviewer: ```suggestion
2. Build and run Detox tests:
```

   
```bash
detox build --configuration ios.sim.release
detox test --configuration ios.sim.release
```
 
Here is how your **codemagic.yaml** should look like:

> ```yaml

Reviewer: check the indentation of the yaml example

  workflows:
    detox-test:

>         name: Deto test automation

Reviewer: ```suggestion
        name: Detox test automation
```

        environment:
          node: latest
          xcode: latest
        scripts:
            - name: Install Detox CLI tools
              script: npm install detox-cli --global
            - name: Install applesimutils
              script: |
                brew tap wix/brew
                brew install applesimutils
            - name: Build Detox app
              script: detox build --configuration ios.sim.release
            - name: Execute Detox testing
              script: detox test --configuration ios.sim.release

> ```

Reviewer: Maybe add a sentence to the end, i.e. "To run this workflow automatically in response to events in the repository, you can additionally configure [automatic build triggering]."
